Update calculations page

chiaberry opened the pull request:
### Related Issue: 
https://github.com/cityofaustin/atd-data-tech/issues/22090

Moving [this code](https://github.com/cityofaustin/atd-knack-signs-markings/blob/staging/src/calcs.js) to typescript. 

Deployed version of existing code https://atd-knack-signs-markings.netlify.app/calcs. Christina said the techs that use this calculations page no longer want the slider/range input, and just number inputs work. 

### How to test
https://deploy-preview-325--nextjs-knack-signs-markings.netlify.app/calcs/

or to run locally: 
check out this branch, make sure you are in the `signs-work-orders-map` directory
`npm install`
`npm run dev`
visit `http://localhost:3000/calcs/`

Did I do the typescript part correctly? I have quieted all of the warnings and errors, but I would love some feedback on if it I did that the right way

### What is left to be done
Adding input validation
Truncating the results to 4? decimal places
Style tweaks, I want to get input from Christina especially about this part

chiaberry committed:
add bootstrap

chiaberry committed:
adding react hook form

chiaberry committed:
messy wip

chiaberry committed:
use the controller to control the inputs

chiaberry committed:
the first card is done

chiaberry committed:
trying to fix the types

chiaberry committed:
adding another one to look for patterns

chiaberry committed:
create the calc component

chiaberry committed:
move the useCalculation into the component

chiaberry committed:
trying to fix the types

chiaberry committed:
remove the slider

chiaberry committed:
add the beads of paint

chiaberry committed:
add optional thickness

chiaberry committed:
is this how to handle if there is no thickness field

chiaberry committed:
move types into own file

chiaberry committed:
defining all of the thermopastic calculations in one array

chiaberry committed:
fix the placeholdre i left in the type

chiaberry committed:
move the paint caluclations into an var

chiaberry committed:
add rpms component

chiaberry commented on signs-work-order-map/components/ControlledNumberInput.tsx:
this originally had the slider and number inputs and the controller made sure they were both in sync, but then last week when I showed Christina she said they didnt use the slider input.

johnclary replied:
was wondering—thanks!

johnclary replied:
could be cool to add big-ole plus and minus buttons to increment/decrement 🧠

johnclary commented on signs-work-order-map/types/calcs.ts:
```diff
@@ -0,0 +1,75 @@
+  calculationName: string;
+}
+
+export interface useCalculationType {
```
i think `UseCalculationProps` would be a bit more clear. And note that you'll definitely want to `TitleCase` all of your type and interface definitions 😄

johnclary commented on signs-work-order-map/app/calcs/calc.css:
```diff
@@ -0,0 +1,6 @@
+    font-weight: 400;
+    font-size: 40px;
+    color: #212529;
+  }
```
Might we use inline bootstrap classes for these styles? Otherewise it'd be nice to use `rem` instead of `px` and establish these in the global scss, since we will hopefully reuse them if we add more forms in the future.

frankhereford reviewed (approved):
Chia, thank you for this. I think your typescript looks great, and I can't see anything wrong with it at all. It's very explicit, well organized, and doesn't produce any warnings at all for me.

If pressed, I might say maybe to define the return types of your functions, but that is a reach, at best. TS is doing just fine inferring them and getting it right.

This is fantastic, thank you. I look forward to learning where this goes in terms of styling and UI, and one idea you might toy with is the use of keyboard shortcuts to jump to certain input fields? In any case, this is already past MVP, IMHO. 🚢 🚢🚢

roseeichelmann committed:
test netlify deploy

johnclary commented on signs-work-order-map/components/ControlledNumberInput.tsx:
```diff
@@ -0,0 +1,37 @@
+  label: string;
+}
+
+export default function ControlledNumberInput({
```
Would be nice to add a little docstring above this that describes what the component does

johnclary commented on signs-work-order-map/components/ControlledNumberInput.tsx:
```diff
@@ -0,0 +1,37 @@
+import { Controller, Control } from "react-hook-form";
+import { CalculationInputsType } from "@/types/calcs";
+
+interface ControlledNumberInputProps {
```
Might be nice to add little docstrings in these so that we get the docs in our type hints 😇

johnclary commented on signs-work-order-map/components/ControlledNumberInput.tsx:
```diff
@@ -0,0 +1,37 @@
+      render={({ field: { onChange, value } }) => (
+        <div className="d-flex justify-content-between">
+          <Form.Label className="col-3">{label}</Form.Label>
+          <Form.Control
```
This component takes a `size` prop that can be set to `"lg"`. might be nice for touch devices?

johnclary commented on signs-work-order-map/types/calcs.ts:
```diff
@@ -0,0 +1,75 @@
+  adhesiveOutput: number;
+};
+
+export interface CalcComponentTypeDef {
```
Would be nice to add docstrings to this guy

johnclary commented on signs-work-order-map/components/ControlledNumberInput.tsx:
```diff
@@ -0,0 +1,37 @@
+        <div className="d-flex justify-content-between">
+          <Form.Label className="col-3">{label}</Form.Label>
+          <Form.Control
+            type="number"
```
We should consider using the `inputmode="numeric` [pattern](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/) here. I'm not sure how relevant these issues still are, but we went that route in the VZE.

johnclary commented on signs-work-order-map/config/calcs.ts:
```diff
@@ -0,0 +1,110 @@
+
+export const thermoplasticCalculations: Array<CalcComponentTypeDef> = [
+  {
+    title: "60 mils thick (maintenance)",
```
i know these are legacy names we're porting over, but it seems like they could be cleaned up.

* **60mm thick**
* **90mm thick**
* **Beads**
* **Thermoplastic Primer**

...?

chiaberry replied:
@ChristinaTremel wondering what you think about reworking some of these titles?

johnclary commented on signs-work-order-map/components/CalcWidthLengthComponent.tsx:
```diff
@@ -0,0 +1,90 @@
+        <h4>{title}</h4>
+      </Row>
+      <Row>
+        <Col>
```
i don't know how these different forms are used, but one way to clean this up visually would be to wrap each section in a `Card` component and move the title into the `Card.Title`. The output going into the `Card.Footer` could help as well 🤔

mddilley replied:
+1 - one thing that stuck with me from the ACC folks was them defaulting to cards for visual separation.

johnclary commented on signs-work-order-map/config/calcs.ts:
```diff
@@ -0,0 +1,110 @@
+export const MAXTHICKNESS = 15;
+export const MAXRPMS = 100;
+
+export const defaultDivisors: { [key: string]: number } = {
```
i am wondering if these can go into the `thermoplasticCalculations` array instead of  a separate config?

johnclary commented on signs-work-order-map/types/calcs.ts:
```diff
@@ -0,0 +1,75 @@
+  outputFieldName: keyof CalculationInputsType;
+}
+
+export interface useSimpleCalculationType {
```
Another casing issue

johnclary commented on signs-work-order-map/components/CalcRPMSComponent.tsx:
```diff
@@ -0,0 +1,62 @@
+    calculationName: calculationName,
+    setValue,
+    outputFieldName: outputName,
+  });
```
This is working fine, but I'm finding it a bit off that you're putting the output calculation into a field tracked by RHF. I wonder if you might simplify this by having this hook wrap a `useMemo` that just returns the output of the math formula—and render that without fussing with `setValue` and `useEffect`. 

And to be honest this math is so trivial that even I am not worried about memoizing.

johnclary commented on signs-work-order-map/app/calcs/page.tsx:
```diff
@@ -0,0 +1,87 @@
+  });
+
+  return (
+    <div>
```
would be good to swap this out with a `Container` component.

johnclary commented on signs-work-order-map/app/calcs/page.tsx:
```diff
@@ -0,0 +1,87 @@
+};
+
+export default function Calcs() {
+  const { control, watch, setValue } = useForm<CalculationInputsType>({
```
i've been thinking about how you might simplify the configs, and one option could be to move the `<form>`/`useForm` hook into the `CalcWidthLengthComponent`. Since these different sections don't rely on each other in any way. I think that'd let you use the same field names for each instance of the form and not have to track so many variables and what have you. Just a suggestion based on my quick scan of the code, so if that sounds appealing, great, otherwise please ignore 👍

mddilley replied:
i like this suggestion! I think that breaking up the one large form into smaller forms that use RHF in both `CalcWidthLengthComponent` and `CalcRPMSComponent` components could make it more straightforward to track where default values and future validation rules are being used without needing to tie the grouped `defaultValues` to the config coming from elsewhere.

i _think_ that could open the door to grouping the default values and other RHF-specific props into the config objects that live in `calc.ts` along with `title`, `inputWidthName`, and the others. Could even allow full customization of each iteration of the form if that ever became important 🤷

edit: finishing review and seeing that this kind of change could force you to break up `CalculationInputsType` and figure out the dynamic types.. so maybe this would be more work than it is worth

johnclary reviewed (commented):
@chiaberry this is looking good. Very nice to see this part of the app come to life. I love the use of react-hook form.

I didn't offer much advice, but I'd like to see the design of this page get polished. I'd be happy to help brainstorm ideas. The font sizing, weights, and labels need attention IMO.

I also really think we should be testing this form on the actual tablet that the techs will be using—and we should test it outdoors in full sunlight.

chiaberry commented:
@johnclary thank you for your review! I will start addressing the feedback

and yes, fully agree that the styling is not done, I have that and input validation as part of next steps.

chiaberry committed:
Merge branch '21517-app-next' into 22090-add-calc-page

chiaberry commented:
@ChristinaTremel this is ready for you to review! 

https://deploy-preview-325--nextjs-knack-signs-markings.netlify.app/calcs/
It is far from polished, I want to check with you if you want the UI to look like the existing UI (minus the slider input), or if you want any other changes.

chiaberry committed:
move the divisors const into the config

chiaberry committed:
simply by removing useeffect and setvalue

chiaberry committed:
change input type to numeric

chiaberry committed:
remove css import and update styles

chiaberry committed:
adding some docstrings

chiaberry committed:
one more note

roseeichelmann reviewed (approved):
the typescript looks good to me, good to know that the techs dont want the slider anymore. i liked johns suggestion to have buttons to increase the inputs but if the techs wouldnt use that and prefer to type in then i dont see the need to add that.

mddilley commented on signs-work-order-map/components/CalcRPMSComponent.tsx:
```diff
@@ -0,0 +1,47 @@
+      </Row>
+      <Row>
+        <Col>
+          <ControlledNumberInput
```
love these reusable input components!

mddilley reviewed (approved):
Had a great time reviewing this one, and it is cool to see RHF in here and bootstrap too. Can't wait to see the next steps! ✨

For validation, it could be worth exploring [zod + RHF resolver](https://github.com/react-hook-form/resolvers?tab=readme-ov-file#zod). I know we have some nice validation examples recently added in VZE too, though.

ChristinaTremel commented:
Love the look of the redesigned page and seems to be functioning exactly how we want it to. 

When I go to the office this week, I'll check out the look and feel of the page on a tablet. I was also thinking I could stop by SMD and see if I can gather some feedback from the Supervisors/a couple Technicians because (if I'm remembering right) I think the "Width" fields only have a handful of options that they would install (_4", 6", 8", 12", and 24" maybe_). If that's the case - I'm wondering if there's any design features we could implement that allow them to choose between those and then type in the linear feet? Or if changing the design for that would be too much, I won't suggest it! I'm happy with the form and I think it's simplified by removing the slider feature they weren't using. 🙂

chiaberry committed:
add card components

mateoclarke commented:
Looks great! I liked seeing RHF to modernize this and the type definitions definitely help to jump into complicated components. 

If we wanted to be extra super careful that all the calculations translated over as expected without introducing new bugs, we could do a few tests, manual or code-based, between the new and old version. This would make sure a few test calculations are consistent with real world numbers. Maybe this is something we could do with users too. All of that is more work, but I leave it to y'alls discretion. 

![calc](https://github.com/user-attachments/assets/93d39733-6e18-455b-9880-b0b3a02d6a53)

mateoclarke reviewed (approved):
🧮

ChristinaTremel commented:
I was able to connect with the SMD team and they really liked the new design functionality! :) I asked them about limiting the width to the 5 options I noted before, but if feels that there are occasional one offs of different widths so I think it's best to keep it as the open field. They did note that we could restrict the width max from going beyond 25 because the widest width they install is 24".

chiaberry commented:
I opened https://github.com/cityofaustin/atd-data-tech/issues/22896 to capture follow up work, merging this down into collector branch.
